src/HOCON/Impl/HoconTokenizer.cs

```diff
@@ -152,62 +65,113 @@ public string Take(int length)
-        public bool Matches(params string[] patterns)
+        protected bool Matches(params string[] patterns)
         {
             foreach (string pattern in patterns)
             {
-                if (pattern.Length + _index >= _text.Length)
+                if (pattern.Length + Index > _text.Length)
                     continue;
 
-                if (_text.Substring(_index, pattern.Length) == pattern)
+                if (_text.Substring(Index, pattern.Length) == pattern)
                     return true;
             }
             return false;
         }
 
-        /// <summary>
-        /// Retrieves the next character in the tokenizer without advancing its position.
-        /// </summary>
-        /// <returns>The character at the tokenizer's current position.</returns>
-        public char Peek()
+        protected bool Matches(char pattern)
         {
             if (EoF)
-                return (char) 0;
+                return false;
+
+            return _text[Index] == pattern;
+        }
+
+        protected bool Matches(params char[] patterns)
+        {
+            if (EoF)
+                return false;
+
+            return patterns.Any(pattern => _text[Index] == pattern);
+        }
+
+        protected bool BackMatch(string pattern)
+        {
+            if (Index - pattern.Length < 0)
+                return false;
 
-            return _text[_index];
+            return _text.Substring(Index = pattern.Length, pattern.Length) == pattern;
         }
 
         /// <summary>
         /// Retrieves the next character in the tokenizer.
         /// </summary>
         /// <returns>The character at the tokenizer's current position.</returns>
-        public char Take()
+        protected char Take()
         {
             if (EoF)
-                return (char) 0;
+                return (char)0;
+
+            var c = _text[Index];
+            Index++;
+            LinePosition++;
+            if (c == Utils.NewLine)
+            {
+                LineNumber++;
+                LinePosition = 1;
+            }
 
-            return _text[_index++];
+            return c;
         }
 
         /// <summary>
-        /// Advances the tokenizer to the next non-whitespace character.
+        /// Retrieves a string of the given length from the current position of the tokenizer.
         /// </summary>
-        public void PullWhitespace()
+        /// <param name="length">The length of the string to return.</param>
+        /// <returns>
+        /// The string of the given length. If the length exceeds where the
+        /// current index is located, then null is returned.
+        /// </returns>
+        protected string Take(int length)
         {
-            while (!EoF && Whitespaces.Contains(Peek()))
+            if (Index + length > _text.Length)
+                return null;
+
+            string s = _text.Substring(Index, length);
```

Comment:
Again, no need to allocate. The current state of HOCON parser can allocate 1000x more memory than the string which it parses. I think, if we are going to rewrite it we also should think about that.

Reply:
Done, I've removed as much `Substring` as possible.